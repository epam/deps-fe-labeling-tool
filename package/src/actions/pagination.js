import { createAction } from 'redux-actions'

const FEATURE_NAME = 'PAGINATION'

const openPage = createAction(`${FEATURE_NAME}/OPEN_PAGE`)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  openPage
}
