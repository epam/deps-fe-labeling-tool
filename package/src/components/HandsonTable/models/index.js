
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
