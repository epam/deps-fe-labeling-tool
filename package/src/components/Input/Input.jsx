import React from 'react'
import { Input as AntdInput } from 'antd'
import PropTypes from 'prop-types'
import 'antd/lib/input/style/index.less'

const Input = ({ innerRef, ...rest }) => (
  <AntdInput {...rest} ref={innerRef} />
)

Input.TextArea = AntdInput.TextArea

Input.propTypes = {
  innerRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({
      current: PropTypes.instanceOf(Element)
    })
  ])
}

export { Input }
