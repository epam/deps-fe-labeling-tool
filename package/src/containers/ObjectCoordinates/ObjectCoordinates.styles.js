import styled from 'styled-components'
import { COLORS } from '@/theme/theme.default'

const InputPrefix = styled.div`
  display: flex;
  justify-content: space-between;
  width: 3rem;
`

const Param = styled.span`
  text-transform: uppercase;
  color: ${COLORS.GRAYSCALE_5};
`

const InputsWrapper = styled.div`
  width: 80%;
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-column-gap: 4.8rem;
  grid-row-gap: 0.8rem;

  > span {
    width: 8.8rem;
  }

  .ant-input-prefix {
    margin: 0;
  }

  .ant-input-affix-wrapper {
    height: 2.4rem;
    padding: 0 0.6rem;
  }
`

export {
  InputPrefix,
  Param,
  InputsWrapper
}
