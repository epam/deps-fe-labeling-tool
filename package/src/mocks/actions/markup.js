const addLabels = jest.fn((labels) => ({
  type: 'MOCK_ADD_LABELS_ACTION',
  payload: labels
}))

const addTables = jest.fn((tables) => ({
  type: 'MOCK_ADD_TABLES_ACTION',
  payload: tables
}))

const removeLabels = jest.fn((labels) => ({
  type: 'MOCK_REMOVE_LABELS_ACTION',
  payload: labels
}))

const updateLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_LABELS_ACTION',
  payload: labels
}))

const updateAllLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_ALL_LABELS_ACTION',
  payload: labels
}))

const updateTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_TABLES_ACTION',
  payload: tables
}))

const updateAllTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_ALL_TABLES_ACTION',
  payload: tables
}))

const selectLabels = jest.fn((labels) => ({
  type: 'MOCK_SELECT_LABELS_ACTION',
  payload: labels
}))

const selectTables = jest.fn((tables) => ({
  type: 'MOCK_SELECT_TABLES_ACTION',
  payload: tables
}))

const clearSelection = jest.fn(() => ({
  type: 'MOCK_CLEAR_SELECTION_ACTION'
}))

const removeTables = jest.fn((tables) => ({
  type: 'MOCK_REMOVE_TABLES_ACTION',
  payload: tables
}))

const importMarkup = jest.fn(() => ({
  type: 'MOCK_IMPORT_MARKUP'
}))

const exportMarkup = jest.fn(() => ({
  type: 'MOCK_EXPORT_MARKUP'
}))

const undo = jest.fn(() => ({
  type: 'MOCK_UNDO_ACTION'
}))

const redo = jest.fn(() => ({
  type: 'MOCK_REDO_ACTION'
}))

const updateInitialMarkup = jest.fn(() => ({
  type: 'MOCK_UPDATE_INITIAL_MARKUP_ACTION'
}))

const storeMarkup = jest.fn((markup) => ({
  type: 'MOCK_STORE_MARKUP_ACTION',
  payload: markup
}))

const storeAssignedMarkup = jest.fn((markup) => ({
  type: 'MOCK_UPDATE_MARKUP_ACTION',
  payload: markup
}))

const updatePageMarkup = jest.fn(() => ({
  type: 'UPDATE_PAGE_MARKUP'
}))

const reset = jest.fn(() => ({
  type: 'RESET_MARKUP_ACTION'
}))

const copyMarkup = jest.fn(() => ({
  type: 'COPY_MARKUP_ACTION'
}))

const pasteMarkup = jest.fn(() => ({
  type: 'PASTE_MARKUP_ACTION'
}))

const updateLabelsWithSettings = jest.fn(() => ({
  type: 'UPDATE_LABELS_WITH_SETTINGS'
}))

const updateTablesWithSettings = jest.fn(() => ({
  type: 'UPDATE_TABLES_WITH_SETTINGS'
}))

const mockMarkupActions = {
  addLabels,
  addTables,
  copyMarkup,
  pasteMarkup,
  removeLabels,
  reset,
  updateLabels,
  updateAllLabels,
  updateTables,
  updateAllTables,
  selectLabels,
  selectTables,
  clearSelection,
  removeTables,
  importMarkup,
  exportMarkup,
  undo,
  updateInitialMarkup,
  storeMarkup,
  storeAssignedMarkup,
  updatePageMarkup,
  redo,
  updateLabelsWithSettings,
  updateTablesWithSettings
}

export {
  mockMarkupActions
}
