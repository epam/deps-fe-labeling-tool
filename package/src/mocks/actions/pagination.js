const openPage = jest.fn((page) => ({
  type: 'MOCK_OPEN_PAGE_ACTION',
  payload: page
}))

const mockPaginationActions = {
  openPage
}

export {
  mockPaginationActions
}
