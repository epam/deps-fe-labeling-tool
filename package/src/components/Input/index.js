import { FilterInput } from './FilterInput'
import { Input } from './Input'
import { NumericInput } from './NumericInput'

Input.Numeric = NumericInput
Input.Filter = FilterInput

export {
  Input
}
