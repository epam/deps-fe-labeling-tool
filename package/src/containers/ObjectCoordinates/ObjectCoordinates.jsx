import React, { useState, useEffect, useMemo, useCallback, useRef } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { updateLabels, updateTables } from '@/actions/markup'
import { Input } from '@/components/Input'
import { useComponentWillUnmount } from '@/hooks/useComponentWillUnmount'
import { usePrevious } from '@/hooks/usePrevious'
import { Label, labelShape } from '@/models/Label'
import * as Table from '@/models/Table'
import { tableShape } from '@/models/Table'
import { getPosition } from '@/models/Table/getPosition'
import { getSize } from '@/models/Table/getSize'
import { pageSelectedLabelsSelector, pageSelectedTablesSelector } from '@/selectors/markup'
import { currentPageSelector } from '@/selectors/pagination'
import { InputPrefix, Param, InputsWrapper } from './ObjectCoordinates.styles'

const INPUT_LENGTH = 6
const CoordSymbol = {
  X: 'x',
  Y: 'y',
  W: 'w',
  H: 'h'
}
const TABLE_TYPE_NAME = 'table'

const coordsToSlicedString = (coords) => Object.fromEntries(
  Object.entries(coords).map(([param, value]) => (
    [param, String(value).slice(0, 6)]
  ))
)

const getInitialCoords = (selected) => {
  if (selected.typeName === TABLE_TYPE_NAME) {
    return coordsToSlicedString({
      ...getPosition(selected),
      ...getSize(selected)
    })
  }

  return coordsToSlicedString(Label.toRectangle(selected))
}

const ObjectCoordinates = ({
  currentPage,
  selectedLabels: [selectedLabel],
  selectedTables: [selectedTable],
  updateLabels,
  updateTables
}) => {
  const selectedObject = selectedLabel ?? selectedTable

  const prevSelectedObject = usePrevious(selectedObject)

  const [coords, setCoords] = useState(getInitialCoords(selectedObject))

  const lastParam = useRef()

  const updateObject = useCallback((param) => {
    if (!param) {
      return
    }

    const numericValue = Number(coords[param])

    if (selectedTable) {
      const tableWithNewPos = param === CoordSymbol.X
        ? Table.setPosition(selectedTable, { x: numericValue })
        : Table.setPosition(selectedTable, { y: numericValue })
      return updateTables(currentPage, [tableWithNewPos])
    }

    return updateLabels(currentPage, [{
      ...selectedLabel,
      [param]: numericValue
    }])
  }, [
    coords,
    currentPage,
    selectedLabel,
    selectedTable,
    updateLabels,
    updateTables
  ])

  useComponentWillUnmount(() => updateObject(lastParam.current))

  useEffect(() => {
    if (prevSelectedObject !== selectedObject) {
      setCoords(getInitialCoords(selectedObject))
    }
  }, [
    coords,
    prevSelectedObject,
    selectedObject,
    selectedLabel,
    selectedTable,
    updateObject
  ])

  const onChangeCoord = (value, param) => {
    lastParam.current = param
    setCoords((prevCoords) => ({
      ...prevCoords,
      [param]: `${value}`
    }))
  }

  const isInputDisabled = useCallback(
    (param) => !!selectedTable && (param === CoordSymbol.W || param === CoordSymbol.H),
    [selectedTable]
  )

  const getCoordPrefix = (prefix) => (
    <InputPrefix>
      <Param>{prefix}</Param>
    </InputPrefix>
  )

  const renderInputs = useMemo(() => (
    Object.keys(coords).map((param) => (
      <Input
        key={param}
        disabled={isInputDisabled(param)}
        maxLength={INPUT_LENGTH}
        onBlur={() => updateObject(param)}
        onChange={(e) => onChangeCoord(e.target.value, param)}
        onPressEnter={() => updateObject(param)}
        prefix={getCoordPrefix(param)}
        value={coords[param]}
      />
    ))
  ), [coords, isInputDisabled, updateObject])

  return (
    <InputsWrapper>
      {renderInputs}
    </InputsWrapper>
  )
}

ObjectCoordinates.propTypes = {
  currentPage: PropTypes.number.isRequired,
  updateLabels: PropTypes.func.isRequired,
  updateTables: PropTypes.func.isRequired,
  selectedLabels: PropTypes.arrayOf(labelShape).isRequired,
  selectedTables: PropTypes.arrayOf(tableShape).isRequired
}

const mapStateToProps = (state) => ({
  currentPage: currentPageSelector(state),
  selectedLabels: pageSelectedLabelsSelector(state),
  selectedTables: pageSelectedTablesSelector(state)
})

const mapDispatchToProps = {
  updateLabels,
  updateTables
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(ObjectCoordinates)

export {
  ConnectedComponent as ObjectCoordinates
}
