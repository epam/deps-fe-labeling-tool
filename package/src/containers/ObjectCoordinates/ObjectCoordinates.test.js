import { mockMarkupActions } from '@/mocks/actions/markup'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockMarkupSelectors } from '@/mocks/selectors/markup'
import { mockPaginationSelectors } from '@/mocks/selectors/pagination'
import React from 'react'
import { shallow } from 'enzyme'
import { updateTables, updateLabels } from '@/actions/markup'
import { Input } from '@/components/Input'
import { Table } from '@/models/Table'
import { ObjectCoordinates } from './ObjectCoordinates'

jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/actions/markup', () => mockMarkupActions)
jest.mock('@/selectors/markup', () => mockMarkupSelectors)
jest.mock('@/selectors/pagination', () => mockPaginationSelectors)

const {
  WrappedComponent,
  mapStateToProps,
  mapDispatchToProps
} = ObjectCoordinates

describe('Container: ObjectCoordinates', () => {
  describe('mapStateToProps', () => {
    it('should pass expected state as props to WrappedComponent', () => {
      const { props } = mapStateToProps()
      expect(props).toMatchSnapshot()
    })
  })

  describe('mapDispatchToProps', () => {
    const { props } = mapDispatchToProps()

    it('should pass updateTables action as prop to WrappedComponent', async () => {
      await props.updateTables([])
      expect(updateTables).nthCalledWith(1, [])
    })

    it('should pass updateLabels action as prop to WrappedComponent', () => {
      props.updateLabels([])
      expect(updateLabels).nthCalledWith(1, [])
    })
  })

  let wrapper, defaultProps

  const event = {
    target: {
      value: '0.3'
    }
  }

  beforeEach(() => {
    defaultProps = {
      ...mapStateToProps().props,
      ...mapDispatchToProps().props
    }

    wrapper = shallow(<WrappedComponent {...defaultProps} />)
  })

  it('should render layout correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should change value in input when triggered onChangeCoord', () => {
    let input = wrapper.find(Input).at(0)
    input.props().onChange(event)
    input = wrapper.find(Input).at(0)
    expect(input.props().value).toEqual(event.target.value)
  })

  it('should call updateLabels with correct args when entering the data', () => {
    const input = wrapper.find(Input).at(0)
    input.props().onPressEnter(event)
    expect(updateLabels).nthCalledWith(1, defaultProps.currentPage, defaultProps.selectedLabels)
  })

  it('should call updateTables with correct args when entering the data', () => {
    defaultProps.selectedLabels = []
    defaultProps.selectedTables = [new Table([0.1, 0.33], [0.32, 0.55])]
    wrapper = shallow(<WrappedComponent {...defaultProps} />)

    const input = wrapper.find(Input).at(0)
    input.props().onPressEnter(event)
    expect(updateTables).nthCalledWith(1, defaultProps.currentPage, defaultProps.selectedTables)
  })
})
