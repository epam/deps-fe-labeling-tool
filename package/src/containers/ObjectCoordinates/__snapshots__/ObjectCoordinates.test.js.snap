// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: ObjectCoordinates mapStateToProps should pass expected state as props to WrappedComponent 1`] = `
Object {
  "currentPage": 1,
  "selectedLabels": Array [
    Label {
      "confidence": 0.8,
      "content": "MOCK FIELD KEY 1",
      "fieldCode": "MOCK_FIELD_1",
      "h": 0.4,
      "index": 0,
      "meta": Object {},
      "type": "key",
      "typeName": "label",
      "uid": "1",
      "w": 0.3,
      "x": 0.1,
      "y": 0.2,
    },
  ],
  "selectedTables": Array [],
}
`;

exports[`Container: ObjectCoordinates should render layout correctly 1`] = `
<ObjectCoordinatesstyles__InputsWrapper>
  <Input
    disabled={false}
    key="x"
    maxLength={6}
    onBlur={[Function]}
    onChange={[Function]}
    onPressEnter={[Function]}
    prefix={
      <ObjectCoordinatesstyles__InputPrefix>
        <ObjectCoordinatesstyles__Param>
          x
        </ObjectCoordinatesstyles__Param>
      </ObjectCoordinatesstyles__InputPrefix>
    }
    value="0.1"
  />
  <Input
    disabled={false}
    key="y"
    maxLength={6}
    onBlur={[Function]}
    onChange={[Function]}
    onPressEnter={[Function]}
    prefix={
      <ObjectCoordinatesstyles__InputPrefix>
        <ObjectCoordinatesstyles__Param>
          y
        </ObjectCoordinatesstyles__Param>
      </ObjectCoordinatesstyles__InputPrefix>
    }
    value="0.2"
  />
  <Input
    disabled={false}
    key="w"
    maxLength={6}
    onBlur={[Function]}
    onChange={[Function]}
    onPressEnter={[Function]}
    prefix={
      <ObjectCoordinatesstyles__InputPrefix>
        <ObjectCoordinatesstyles__Param>
          w
        </ObjectCoordinatesstyles__Param>
      </ObjectCoordinatesstyles__InputPrefix>
    }
    value="0.3"
  />
  <Input
    disabled={false}
    key="h"
    maxLength={6}
    onBlur={[Function]}
    onChange={[Function]}
    onPressEnter={[Function]}
    prefix={
      <ObjectCoordinatesstyles__InputPrefix>
        <ObjectCoordinatesstyles__Param>
          h
        </ObjectCoordinatesstyles__Param>
      </ObjectCoordinatesstyles__InputPrefix>
    }
    value="0.4"
  />
</ObjectCoordinatesstyles__InputsWrapper>
`;
